clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp

```diff
@@ -1,28 +1,41 @@
-// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes -emit-llvm -o - %s | FileCheck %s
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
 
 class Foo {
 public:
   [[intel::no_global_work_offset(1)]] void operator()() const {}
 };
 
-template <typename name, typename Func>
-__attribute__((sycl_kernel)) void kernel(const Func &kernelFunc) {
-  kernelFunc();
-}
+template <int SIZE>
+class Functor {
+public:
+  [[intel::no_global_work_offset(SIZE)]] void operator()() const {}
+};
+
+int main() {
+  q.submit([&](handler &h) {
+    Foo boo;
+    h.single_task<class kernel_name1>(boo);
 
-void bar() {
-  Foo boo;
-  kernel<class kernel_name1>(boo);
+    h.single_task<class kernel_name2>(
+        []() [[intel::no_global_work_offset(1)]]{});
 
-  kernel<class kernel_name2>(
-      []() [[intel::no_global_work_offset]]{});
+    h.single_task<class kernel_name3>(
+        []() [[intel::no_global_work_offset(0)]]{});
 
-  kernel<class kernel_name3>(
-      []() [[intel::no_global_work_offset(0)]]{});
+    Functor<1> f;
+    h.single_task<class kernel_name4>(f);
+  });
+  return 0;
 }
 
-// CHECK: define spir_kernel void @{{.*}}kernel_name1() {{.*}} !no_global_work_offset ![[NUM5:[0-9]+]]
-// CHECK: define spir_kernel void @{{.*}}kernel_name2() {{.*}} !no_global_work_offset ![[NUM5]]
-// CHECK: define spir_kernel void @{{.*}}kernel_name3() {{.*}} ![[NUM4:[0-9]+]]
+// CHECK: define spir_kernel void @{{.*}}kernel_name1"() #0 {{.*}} !no_global_work_offset ![[NUM5:[0-9]+]]
+// CHECK: define spir_kernel void @{{.*}}kernel_name2"() #0 {{.*}} !no_global_work_offset ![[NUM5]]
+// CHECK: define spir_kernel void @{{.*}}kernel_name3"() #0 {{.*}} ![[NUM4:[0-9]+]]
+// CHECK: define spir_kernel void @{{.*}}kernel_name4"() #0 {{.*}} !no_global_work_offset ![[NUM5]]
 // CHECK-NOT: ![[NUM4]]  = !{i32 0}
-// CHECK: ![[NUM5]] = !{}
+// CHECK: ![[NUM5]] = !{i32 1}
```

Comment:
I think the attribute works in a bit another way now. What was before: if value of "Enabled" parameter was true, then metadata no_global_work_offset was added. Now it looks like it can be added even if attribute argument was "0" but with different number in corresponding metadata. I can't see it in the test, but code looks like it does so. Am I wrong? Anyway, I can clearly see that change - ![[NUM5]] = !{i32 1}, are we supposed change LLVM IR output in this patch?

Reply:
@Fznamznon thanks for looking into the patch. Yes, you are right that current support breaks the existing codes. Codegen seems working wrong way. Codegen part needs to redesign here with the template support. Any thought?

>>are we supposed change LLVM IR output in this patch?
Probably not.

Reply:
This is fixed now.